### DISCUSSION TITLE
NIST data sets for model performance testing

### DISCUSSION THREAD

**Discussion by bkoconnell** [bkoconnell/datascience-capstone#31] [General Discussions & Guidelines]
@kristinpredeck @Cadamson-prog 

After some tuning/optimization, we'll be at that point of wanting to test our models outside of the NFI bubble. I had previously done a deep-dive on the NIST data set so I'm going to post some relevant notes on this discussion board for your awareness, and certainly if you have any additional findings that I missed or I misinterpreted please don't hesitate to share. 

Appendix D from the EDA report is a bit of a rant/blurb that I put together. I'll share that in a follow up post because it's worth understanding the nuances of the data files, quant methods, label choices, etc.

First --- I pulled all the pertinent raw data files into our repo:
https://github.com/bkoconnell/datascience-capstone/tree/main/data/raw/NIST

Each source sample has its own subdirectory, and each includes the "configuration.txt" file (see my appendix notes on that). The appendix also details how the parquet files were compiled from the HDZ + PXZ files, and it also explains why there are so many variations of the neQuant data sets, so I won't go into that detail here but I recommend reading up on it.

One important note that I will mention --- the "data.parquet" file is the real-time processing data, but it only processed against 17 reference standard spectra which did not include lead or antimony. So, don't bother using it. It's there for reference. The other parquet files are the reprocessed files which include a wider range of elements including the core GSR elements. 

We've previously confirmed that all elements in the NIST dataset exist in the NFI dataset. Feel free to double check that if you'd like.

If you're only interested in shooter samples, any of the sources that begin with "shooter" will have some GSR particles and you'll have a nice selection of different quant methods and label systems to choose from. 

Okay so now, instead of having to go through each sample individually, if you go to the data/processed/ directory you'll see I've concatenated several of the quant types across all samples:
https://github.com/bkoconnell/datascience-capstone/tree/main/data/processed/nist_concatenated_parquets

neQuant and its C variants were the only methods that existed in 100% of the samples, so that's why they're the only ones that I concatenated. 

In these parquets, there are a bunch of metadata files that can be dropped (related to the imaging hardware readings, etc.) and then for each raw element there's also a column that starts with a U_ which is a weighted "uncertainty" column. I haven't done much with this, may make sense to drop it so it's not accidently being read as a raw element. 

The "CLASS" column is the label, and I think the classification is based on the percentages of those "FIRSTELM, SECONDELM, THIRDELM, FOURTHELM" columns which are the heaviest elements in that particle.  I have seen instances where these columns are not accurately displaying either the element name or the weight value, it's hard to tell which, but I believe these columns should be dropped because they feel like they're just redundant noise possibly used for originally classifying the particle.

I've added a "sample source" column so we know which source the particle came from. Also notice I've added to "sum" columns at the end related to oxygen. This is a noteworthy difference between NIST and NFI datasets, which I mention in the appendix, but essentially the NFI data has its elemental weights from 0-100 which includes oxygen, BUT the NIST data removes oxygen from the 0-100 weight. Notice that the "sum without oxygen" is 100 (sum of all elements in the particle except for oxygen) and then the "sum with oxygen" is over 100, presumably 100 + the derived weight of the oxygen element. I'm not sure if this will even have any impact on the non-linear models but it's certainly worth noting. In any case, these columns can be dropped, they're just for reference.

Okay so my last thought on this chaotic rant lol --- if you take one of those concatenated NIST datasets and filter out any sample source that begins with "shooter" you will have (in theory) 100% Non-GSR particles. It is all confounders --- the fireworks, the brake pads, etc. Regardless of what is in the "CLASS" column, we know that these source samples were intentionally picked because they're GSR confounders, so really we should expect that all particles are Non-GSR.   Even though when you look at them you'll see classes like GSR.2 or GSR.Pb-Ba --- I don't think these are to be interpreted as "GSR", they're just a classification of the top elements in the particle and I believe this is where our models can/will be magical and will hopefully be able to determine that 100% (or very close) of these confounder particles are Non-GSR.

On the other hand --- analyzing the particles that are only "shooter" might be a bit more challenging. The labeling system is not a 1-for-1 with NFI dataset. It is not a binary as "GSR" vs "Non-GSR". Instead, the CLASS labels are intended to mean "characteristic of GSR", but that is not the same as GSR. So there may be a lot of assumptions if we go into the shooter samples and start deciding what is GSR and what is Non-GSR.

Perhaps what would be most useful would be running all the non-shooter data through our models. Then, running all the shooter data through the models. We won't necessarily know the "Ground truth" for GSR vs Non-GSR, other than that the non-shooter particles should be non-gsr, but if we compare the model results between the non-shooter and the shooter we would expect to see virtually no GSR in the non-shooter, whereas the shooter datasets should have significantly more GSR particles.  We wouldn't necessarily be testing for precision, but we'd have an idea if our models are able to generalize to the NIST datasets.

**Comment by bkoconnell** [comment]
### Appendix D: EDA for NIST dataset 

Our primary focus will be on the NFI dataset, but we may end up using the NIST dataset for model testing. After some initial EDA on the NIST dataset, there are some noteworthy observations that may need revisiting if we decide to use it for model testing. 

For starters, all 30 sources have configuration files that have been reviewed. The data collection and general configuration for all 30 samples has been confirmed to be the same. However, aside from the obvious “GSR” vs. “Non-GSR” descriptors, there are some differences in the reprocessing of data which could have implications in terms of model accuracy. There are also some differences in the structure of the data as well as the scaling techniques and potentially quantification methods as compared to the NFI dataset. These differences will be detailed below. 

The main files retrieved from the NIST source are stored as a pair of files known as a Zeppelin, which consists of an HDZ header file and PXZ data file. The steps for processing these files are outlined here: 
[https://github.com/bkoconnell/datascience-capstone/blob/main/data/raw/NIST/README ](https://github.com/bkoconnell/datascience-capstone/blob/main/data/raw/NIST/README.md)[.md](https://github.com/bkoconnell/datascience-capstone/blob/main/data/raw/NIST/README.md) 

Steps include downloading the Julia programming software, installing specific dependencies, and running some commands that can be found in our scripts directory in the GitHub repository. A batch script was also created to process all raw data files into parquet files for EDA. These parquet files can be found in the data/raw/ directory of our GitHub repository. 

The original data from the particle scan is in the data zeppelin, which was processed in real-time using a “QuickQuant” method of quantification, but it only processed the data against 17 reference standard spectra which did not include Pb (lead) or Sb 
(antimony). The converted “data.parquet” files have been reviewed to verify this. Since 2 out of the 3 core GSR elements are not included in the original data files, these will not be of much use for our research. To be clear, Pb and Sb are not missing in the sense that they weren’t in the particle, they’re only missing from the original data processing because the configuration file specifically did not tell the software to look for them (they were not part of the original 17 elements listed in the config file): 
 
 
After the initial processing, the operator reprocessed the TIFF/MSA data against the 25 keV spectra, which does include Pb and Sb.  In total, the reprocessed files include 32 total elements (15 more than the original processing). All 32 of these elements exist in the NFI dataset. The most important thing to recognize about the reprocessed data is that there are different quantification methods used and different classification rules used. 

The following statements are true for all 30 samples: 

- All have data Zeppelins that we cannot use because they did not process any values for Pb or Sb 
- All have reprocessed the original data against the expanded 25 keV spectra which includes Pb and Sb 
- All have used the NeXLParticle vector file which details the quantification method for reprocessing the original data. The reprocessed data is stored in the neQuant Zeppelin, which is capped at 5,000 particle observations and does not include any GSR class labels 
- All have the same C-variant reprocessed neQuant files: 
 -- neQuantC0 (generic class labels) 
             -- neQuantC0s (JORS[GSR+Base] class labels) 
             -- neQuantC3 (generic class labels) 
             -- neQuantC3s (JORS[GSR+Base] class labels) 
             -- neQuantC6s (generic class labels) 
             -- neQuantC6s (JORS[GSR+Base] class labels) 

- C-variant neQuant files contain all 16k+ particle observations (not capped at 5k) 
- The NeXL open-source library suite (used for neQuant reprocessing) uses standards-based k-ratio quantification and iterative XPP matrix correction 
- All neQuant and C-variants include “uncertainty” estimates (element columns prefixed with U_), representing how confident the quantification is for that element in that particle. It is the one-sigma uncertainty estimate in the same unit as the element (weight percentage) 

The shooter sources additionally have some Zeppelin files that used MLLSQ / MLLSQU (multiple linear least squares) quantification methods. 

There is a “RELOCATED” folder which contains particle data identified as GSR. 

Important note regarding Oxygen: 
The NFI dataset includes oxygen in the sum of its elemental weights, so all 89 elements in the NFI particle data will sum to 100%.  In contract, NIST does not include oxygen in its scale. All elements for a particle except for oxygen will add up to 100, then oxygen is added on top of that. The author (Ritchie) may have alluded to this as “oxygen-by-stoichiometry” in an article where he wrote: 

“ If it is not possible or not desirable to measure an element, this element can often be modeled using an unmeasured element rule. An example of one such rule is the oxygen-by-stoichiometry model.” 

This may warrant further investigation. It seems that instead of measuring oxygen, like the other elements were measured, the oxygen value is obtained via calculation based on some assumptions of the other elements. It is unknown whether the difference in oxygen quantification methods between the NFI dataset and the NIST dataset has any impact on a model’s ability to classify GSR. It is suspected that some models may be more sensitive to this, while others may be able to ignore it. 

Lastly, in terms of the full dataset, the GitHub repository for our project now contains the converted Zeppelin files in Parquet format in the data/raw/NIST/ folder for the corresponding sample source. At the very least, each of the 30 sample sources has parquet files for each of the neQuant reprocessed datasets and all of the neQuant C-variants.

For convenience, and possibly future EDA, all neQuant variants also have Parquet formatted files that are fully concatenated across all 30 sample sources. These are located in the data/processed/nist_concatenated_parquets/ folder. Be aware that the concatenated files currently have 2 calculated fields added to them which were used to confirm the notes made regarding oxygen quantities. These 2 new fields are elemental sums with oxygen, and element sums without oxygen. It may be necessary to remove these columns if further EDA or any model testing are to be performed.

**Comment by kristinpredeck** [comment]
@bkoconnell

This is super helpful, thanks for doing the deep dive and getting everything organized in the repo. A few thoughts from the DL side:

On the NIST validation approach. I'm fully on board with your plan. The two-part strategy makes a lot of sense,run non-shooter samples through and expect ~0% GSR (pure specificity test), then run shooter samples and look for a meaningful contrast in GSR classification rates vs. non-shooter. You're right that we won't have clean ground truth on the shooter side, but the relative difference tells us whether the model learned generalizable GSR signatures or just memorized NFI patterns. 

A few things I want to flag from the DL preprocessing side:

1. Oxygen renormalization is going to be important. Our MLP was trained on NFI data where elemental weights include oxygen and sum to ~100%. So before I can run NIST through the model, I need to do some work on renormalizing. Without this, the compositional relationships the CLR transform learned from NFI will be meaningless on NIST data. I'll handle this.

2. Five NFI informative elements are missing from NIST (Na, As, Br, W, Hg). I'll zero-fill these, which is defensible since they're likely below the NIST instrument's detection threshold. The 14 extra NIST elements not in our informative set get dropped.

3. On the FIRSTELM/SECONDELM/etc. columns, agree, dropping them. They're redundant and the labeling feels inconsistent from what I've seen.

4. On the CLASS column for non-shooter samples, I noticed that the non-shooter data has literally zero GSR-related classes (no Pb-Ba, no Ba-Sb, nothing). That's clean. The shooter data has Pb-Ba (264), Ba-Sb (204), Pb-Sb (103), and Char GSR-like (694). So there's real signal to find on the shooter side, which is encouraging.

On DL model plan going forward:

The MLP baseline is at 99.9% accuracy with ROC-AUC 1.00 on NFI test data, so traditional hyperparameter tuning (layer sizes, learning rates) probably won't move the needle. Instead I'm focusing on:

- Threshold optimization since the default 0.5 cutoff isn't forensically optimal. I want to use precision-recall curves to find the operating point that minimizes false positives while keeping recall acceptable. This is the most practically useful "tuning" we can do for stakeholders.

- Monte Carlo Dropout for uncertainty quantification: instead of a single probability, we get a distribution of predictions per particle. High-variance particles (especially the barium confounders) get flagged for expert review instead of a hard classification. This directly addresses the false positive problem.

- TabNet: I mentioned this in the analysis plan as an additional model. Its main value at this point isn't beating MLP on accuracy (unlikely), it's the built-in attention masks that show which features the model focuses on per particle. That's a strong interpretability story for the forensic/courtroom stakeholder angle. I'll fit it and compare, but I want to be upfront that it's more about interpretability than performance lift. Let me know if there's anything else you noticed in the data that I should watch out for.

**Reply by bkoconnell** [reply]
"Five NFI informative elements are missing from NIST (Na, As, Br, W, Hg)"   Ohhh rightt.. I'm thinking back to when we had the set of 89 in NFI, so yeah makes sense, we dropped those lol.


"On the CLASS column for non-shooter samples, I noticed that the non-shooter data has literally zero GSR-related classes " 
^^^ I was seeing this also at first, but I think some sample sources are going to be more problematic than others. For example, if you look at the ford brakes like a213 rear driver, there are a lot of particles labeled GSR.Pb-Ba, some have trace amounts of Sb but most are 0, while the barium is around 50% and the lead is between 5-10%.    It will be great to see how the models interpret this.
